core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
       throw new RuntimeException(e);
     }
 
+    startMasters(false);
     while (!Thread.interrupted()) {
-      startMasters(false);
-      LOG.info("Secondary started");
       mLeaderSelector.waitForState(State.PRIMARY);
+      if (gainPrimacy()) {
+        mLeaderSelector.waitForState(State.SECONDARY);
+        losePrimacy();
+      }
+    }
+  }
+
+  /**
+   * @return true if the master successfully upgrades to primary
+   */
+  private boolean gainPrimacy() throws Exception {
+    AtomicBoolean lostPrimacy = new AtomicBoolean(false);
+    try (Scoped scoped = mLeaderSelector.onStateChange(state -> lostPrimacy.set(true))) {
+      if (mLeaderSelector.getState() != State.PRIMARY) {
+        lostPrimacy.set(true);
+      }
       stopMasters();
       LOG.info("Secondary stopped");
-      mJournalSystem.setMode(Mode.PRIMARY);
-
-      startMasters(true);
-      mServingThread = new Thread(() -> {
-        try {
-          startServing(" (gained leadership)", " (lost leadership)");
-        } catch (Throwable t) {
-          Throwable root = ExceptionUtils.getRootCause(t);
-          if ((root != null && (root instanceof InterruptedException)) || Thread.interrupted()) {
-            return;
-          }
-          ProcessUtils.fatalError(LOG, t, "Exception thrown in main serving thread");
+      mJournalSystem.gainPrimacy();
+      if (lostPrimacy.get()) {
```

Comment:
Why do we need to check `lostPrimacy`, and not `mLeaderSelector.getState()`?

Reply:
1. master A becomes primary and begins journal upgrade
2. master A becomes secondary (continues journal upgrade)
3. master B becomes primary and serves some requests
4. master A becomes primary again
5. master A finishes journal upgrade, checks `mLeaderSelector.getState()` and starts serving, but loses the requests that master B served while A was upgrading.